openghg/store/base/_base.py

```diff
@@ -179,28 +179,49 @@ def add_metakeys(self, force: bool = False) -> dict:
 
         return self.metakeys
 
-    def update_metadata(self, data: Dict, input_parameters: dict, additional_metadata: Dict) -> Dict:
+    def update_metadata(self, data: dict, input_parameters: dict, additional_metadata: dict) -> dict:
         """This adds additional metadata keys to the metadata within the data dictionary.
 
         Args:
-            data: Dictionary containing data and metadata for species
-            input_parameters: Input parameters from read_file.
+            data: Dictionary containing data and metadata for datasource
+            input_parameters: Input parameters from read_file...
             additional_metadata: Keys to add to the metadata dictionary
         Returns:
             dict: data dictionary with metadata keys added
         """
+        from openghg.util import merge_dict
+
         # Get defined metakeys from the config setup
         metakeys = self.add_metakeys()
         required = metakeys["required"]
+        # We might not get any optional keys
+        optional = metakeys.get("optional", {})
 
-        input_parameters_required = {key: value for key, value in input_parameters.items() if key in required}
-
-        # Basic implemntation of this
-        # TODO: Move this somewhere else? This shouldn't need self but does need to understand form of data.
-        # TODO: May want to add checks to make sure we're not overwriting anything important.
         for parsed_data in data.values():
-            parsed_data["metadata"].update(input_parameters_required)
-            parsed_data["metadata"].update(additional_metadata)
+            metadata = parsed_data["metadata"]
+
+            # Sources of additional metadata - order in list is order of preference.
+            sources = [input_parameters, additional_metadata]
+            for source in sources:
+                metadata = merge_dict(metadata, source, keys_dict2=required)
```

Comment:
Should we use resolve mismatch here? I just realised my inline comment above isn't correct, the default is to error on conflicting values. In the past we would have overwritten values silently.

Raising a warning seems reasonable, or an error with an option to pass `resolve_mismatch=True` through. (That would have to be propagated through from `standardise...`. I guess since we can't pass kwargs to `standardise_surface` etc, the flag would need to be added to all of the standardise functions and read file functions.

Reply:
I've updated the logic here to use the "left" value by default and log a warning. The `resolve_mismatch` flag has now been replaced by `on_conflict` with the default being "left".

Reply:
Okay, great, thanks!